zeus/optimizer/batch_size/client.py

```diff
@@ -0,0 +1,185 @@
+
+import httpx
+import pynvml
+
+from zeus.callback import Callback
+from zeus.monitor import ZeusMonitor
+from zeus.optimizer.batch_size.common import (
+    GET_NEXT_BATCH_SIZE_URL,
+    REGISTER_JOB_URL,
+    REPORT_RESULT_URL,
+    JobConfig,
+    JobSpec,
+    ReportResponse,
+    TrainingResult,
+)
+from zeus.optimizer.batch_size.exceptions import (
+    ZeusBSOConfigError,
+    ZeusBSOOperationOrderError,
+    ZeusBSORuntimError,
+    ZeusBSOTrainFailError,
+)
+from zeus.util.logging import get_logger
+
+logger = get_logger(__name__)
+
+
+class BatchSizeOptimizer(Callback):
+    """Batch size optimizer client that talks to server. One batch size optimizer per one training session of the job."""
+
+    def __init__(self, monitor: ZeusMonitor, server_url: str, job: JobSpec) -> None:
+        """Initialize the optimizer, and register the job to the server.
+
+        If job is already registered, check if the job configuration is identical with previously registered config.
+
+        Args:
+            monitor: zeus monitor
+            server_url: url of batch size optimizer server
+            job: job specification. Refer to `JobSpec` for job specifcatio parameters.
+        """
+        self.monitor = monitor
+        self.server_url = server_url
+        self.cur_epoch = 0  # 0-indexed
+        self.running_time = 0.0
+        self.consumed_energy = 0.0
+        self.train_end = False
+
+        # Get max PL
+        pynvml.nvmlInit()
+        pls = []
+        for index in self.monitor.nvml_gpu_indices:
+            device = pynvml.nvmlDeviceGetHandleByIndex(index)
+            # name = pynvml.nvmlDeviceGetName(device)
+            # TODO: CHECK DEVICE ALL EQUAL AND SET JOB.GPU_MODEL
+            pls.append(pynvml.nvmlDeviceGetPowerManagementLimitConstraints(device))
+        if not all(pls[0] == pl for pl in pls):
+            raise ZeusBSOConfigError("Power limits ranges are not uniform across GPUs.")
+
+        # set gpu configurations(max_power, number of gpus, and gpu model)
+        self.job = JobConfig(
+            **job.dict(),
+            max_power=(pls[0][1] // 1000) * len(monitor.gpu_indices),
+            number_of_gpus=len(monitor.gpu_indices),
+        )
+
+        # Track the batch size of current job
+        self.current_batch_size = 0
+
+        # Register job
+        res = httpx.post(self.server_url + REGISTER_JOB_URL, content=self.job.json())
+        self._handle_response(res)
+
+        logger.info("Job is registered: %s", str(self.job))
+
+    def get_batch_size(self) -> int:
+        """Get batch size to use from the BSO server.
+
+        Returns:
+            return a batch size to use for current job
+
+        Raises:
+            `ZeusBSORuntimError`: if the batch size we receive is invalid
+        """
+        if self.train_end is True:
+            # If train is already over, should not re-send the request to the server. Typically, re-launch the script for another training
```

Comment:
Hmm. I haven't thought about error cases. If bso is in `MAB` stage there is no problem since `report` is not tied to the specific issuance of that batch size, so user can just re-launch, server re-issue batch size, client report, and update arm states. If it is in the pruning stage, then it will be a big problem, since the exploration state will be left `exploring` and it will wait for the report. 

If the client errors out, we can report to the server that the client errored out to the server in `__exit__`. Then the server can delete that exploration or whatever. The usage of bso might get a little different since we need to use `__exit__`. 

If the server errors out for some reason, then I am not sure... Any suggestion?

Reply:
Is it possible to make the server-side of `get_batch_size` idempotent? Basically if it's idempotent, calling it back-to-back infinitely many times is fine.

Reply:
Hmm very tricky. Unless the user specifies something like `recurrence_number` there is no way for the server to identify which recurrence this training is in. Currently, server assumes that every call of `get_batch_size` is a start of new training, and givees the corresponding batch size. However, if we cannot assume that, we should be able to know what recurrence it is to give the correct batch size for that recurrence.

Reply:
I guess due to concurrent jobs, making `get_batch_size` idempotent is fundamentally not possible. I can't think of a good solution immediately. Some ideas:
- Assigning a training run a unique ID (probably just increment an integer each time `register_job` or `get_batch_size` is called), and the training run reports back with that run ID. Perhaps this will help distinguish concurrent jobs as well.
- Use the `atexit` hook to notify the server that a job finished (be it with an error or successfully). But this is more of a hack.

Reply:
Probably the second approach is fine. You just create a new endpoint which the `atexit` hook sends, and it that comes before the job reporting its training result, you know it died tragically.

Reply:
Shoot. I think the fundamental problem is the only information we get from the client is (job_id, batch_size), so we cannot tell if it is an exploration or a concurrent job. 

In this scenario, 
1. server issue batch size 16 to client A 
2. server issue batch size 16 to client B(concurrent job) 
3. Client B died, notifying (job_id, batch_size = 16) failed 
4. server will remove exploration state. 
5. Client A reports the result back to the server, since there is no exploration state related to client A, it doesn't update the exploration state and treats it as a concurrent job. 

Two answers I can think of for this scenario.
1. It is fine since client B is expected to re-launch the job and server will issue 16 again since there is no exploration state related to 16. So eventually, we will have the same result. 
2. Modify step 4. We can mark that exploration as `Failed` state. Then, when client A reports the result back, we can update the state based on the result. If client B re-launched the job before client A reports the result, we can just issue 16 again since it's in `Failed` state. So, the result will be the same. 

Another possible solution is switching our DB schema, which is a pretty big change .. haha. I initially thought `recurrence_number` is not necessary, but distinguishing a concurrent job is such a pain without it. 

We change our `ExplorationState` table to `ISSUED` table. In this table, we record all batch_size we issued with recurrence number (something like a unique ID for each run as you mentioned). Then we will have something like
```
Recurrence number | Job_id | batch_size | Type = (MAB or Pruning or Concurrent) 
(Recurrence number & Job_id & batch_size) as a PK
```
Then, we modify `Measurement` table to be the result of each recurrence of training (1-to-1)
```
Recurrence number | Job_id | batch_size | state = (Failed, succeeded) | time | energy | converged.. etc 
State means training errored our or not. 
```
This way, when the client errors out, it can send `job_id,batch_size,recurrence number` combination to update the state. Also, as you mentioned in a different conversation, we can re-construct the whole sequence of issuing batch_ sizes based on recurrence number so its not that hard to re-construct exploration sequence.

Reply:
I actually wrote down something similar but deleted it since thought it would be too much overhead for you. Since it eventually got brought up, I'll say it.

I thought we can merge `ExplorationState` and `Measurement` into one table called `Trial`, which records the complete timeline of all training trials that ever ran and is running.

Fields
- Job ID
- Batch size
- Trial ID (Integer that auto-increments from 1)
- Timestamp (When the trial was created; pretty much when `get_batch_size` was called)
- Trial type (Exploration | Concurrent | MAB)
- Status (Dispatched | Succeeded | Failed)
- Time
- Energy
- Cost
- Converged

When the client asks for the next batch size, the server returns the batch size and its trial ID. The client will report training results & training completion back with its trial ID, so that the server can more easily distinguish what that training trial was by inspecting `trial_type` and update its `status`.

Regarding `recurrence_number`, I feel like it's actually not well-defined when there are concurrent jobs. Recurrence is really assuming that the organization wants to periodically re-train the model in a sequential manner, but concurrent jobs are already breaking that assumption. Say the organization starts two jobs in parallel. What are their recurrence numbers? With that said, let's just forget about `recurrence_number` everywhere. Everything is just trials, and the user is expected to keep retrying until convergence.

Obviously at this point you know way better than me about how measurements, jobs, etc. should be managed and how they interact with other parts of the codebase. Let me know if the `Trial` table works or not. The overall intuition I had is that (1) we don't need to forcefully separate tightly coupled information (exploration state and training results) into separate tables and (2) we can just do a bit more bookkeeping to make our lives easier.

Reply:
I think my definition of `recurrence_number` is the same as your `Trial number`, which represents the how many times it has launched the training regardless of convergence or not. I think `Trial` table makes sense, and I will work on it. I think just refactoring is easier than thinking about tiny corner cases and bugs. Shouldn't take that long time since I know my code.. hopefully haha.

Reply:
Thanks!